### ISSUE TITLE
Hype docker image gives an error

### ISSUE THREAD

**Issue by SarahAlidoost** [eWaterCycle/ewatercycle-hype#5]
The hype notebook using docker image gives the error (the singularity works fine):

```python
--------------------------------------------------------------------------
DeadContainerException                    Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 cfg_file, cfg_dir = model.setup()
      2 cfg_file, cfg_dir

File ~/GitHub/ewatercycle/src/ewatercycle/models/hype.py:152, in Hype.setup(self, start_time, end_time, crit_time, cfg_dir)
    150 # start container
    151 work_dir = str(cfg_dir_as_path)
--> 152 self.bmi = _start_container(self.version, work_dir)
    154 return str(cfg_file), work_dir

File ~/GitHub/ewatercycle/src/ewatercycle/models/hype.py:217, in _start_container(version, work_dir)
    215 elif CFG["container_engine"].lower() == "docker":
    216     image = _version_images[version]["docker"]
--> 217     return BmiClientDocker(
    218         image=image,
    219         image_port=55555,  # TODO needed?
    220         work_dir=work_dir,
    221     )
    222 else:
    223     raise ValueError(
    224         f"Unknown container technology in CFG: {CFG['container_engine']}"
    225     )

File ~/mambaforge/envs/ewatercycle/lib/python3.10/site-packages/grpc4bmi/bmi_client_docker.py:88, in BmiClientDocker.__init__(self, image, work_dir, image_port, host, input_dirs, user, remove, delay, timeout)
     86         logs = self.logs()
     87         msg = f'Failed to start Docker container with image {image}, Container log: {logs}'
---> 88         raise DeadContainerException(msg, exitcode, logs)
     90 super(BmiClientDocker, self).__init__(BmiClient.create_grpc_channel(port=port, host=host), timeout=timeout)

DeadContainerException: Failed to start Docker container with image ewatercycle/hype-grpc4bmi:feb2021, Container log: terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
```

**Comment by BSchilperoort** [comment]
@nishadhka did you get the Hype container working and responding to a grpc4bmi client? 

You can try to connect the grpc4bmiclient to the hype container with these instructions; https://grpc4bmi.readthedocs.io/en/latest/usage.html#python-bmi-client

If the container works and the model responds to grpc4bmi, then it should be possible to get Hype to work again with eWaterCycle. However, it's been a while since it worked so mileage may vary.

@MarkMelotto can you help Nishadh with this?

**Comment by MarkMelotto** [comment]
yep, I can assist with this!

**Comment by nishadhka** [comment]
Thank you @BSchilperoort for the note and @MarkMelotto , have tried on compiling the Dockerfile with few changes and failed, following attempts

Attempt 1: Keeping the Hype old version 
Docker build failed due to the old version of Hype and missing file, bmi_class.h, https://github.com/nishadhka/hype-bmi/commit/1ba45b978a8483493336f7f5e1959f383e1be3a5

Attempt 2: Hype latest version 
Latest version of Hype model but fails due to the issue with updates needed for hypelib.f90, https://github.com/nishadhka/hype-bmi/commit/5597e233a0f1a1b01bed092eb44daf8cd1940bd1

Will check on  grpc4bmiclient

**Comment by BSchilperoort** [comment]
> Docker build failed due to the old version of Hype and missing file, bmi_class.h, [nishadhka/hype-bmi@1ba45b9](https://github.com/nishadhka/hype-bmi/commit/1ba45b978a8483493336f7f5e1959f383e1be3a5)

`bmi_class.h` was available in an older version of grpc4bmi; https://github.com/eWaterCycle/grpc4bmi/tree/v0.2.17/cpp
This was using BMI version 1.0, instead of 2.0.

The grpc4bmi client is backwards compatible so a new client _should_ be able to communicate with an old server.

> Attempt 2: Hype latest version

For this you would indeed have to update the hype-bmi repository to make use of BMI 2.0, which is probably more work. It's probably easier to use the old grpc4bmi.

**Comment by nishadhka** [comment]
Will check and get back. 
The compilation is working,  https://github.com/nishadhka/hype-bmi/commit/4ae463bc1a073fd6dcf77c6966d97cffa88719a9.
But running the image shows a similar error as in https://github.com/eWaterCycle/ewatercycle-hype/issues/5#issuecomment-1993792382, but no core dump reported. Any quick way to test the docker build correctness. 

```
$ sudo docker run -ti --rm hype-bmi-fixed:latest
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string: construction from null is not valid
$
```

**Comment by nishadhka** [comment]
It looks like issue with gRPC4BMI as indicated by Claude. 
```
$ sudo docker run -ti --rm hype-bmi-fixedv4:latest /opt/hype-bmi/src/test_bmi_direct
Starting HYPE BMI server...
Arguments: 2
argv[0]: /opt/hype-bmi/src/hype_bmi_server
argv[1]: /opt/hype-bmi/src/test_bmi_direct
HypeBmi constructor called
Created HypeBmi model
Starting gRPC server with 2 arguments:
  server_argv[0]: /opt/hype-bmi/src/hype_bmi_server
  server_argv[1]: /opt/hype-bmi/src/test_bmi_direct
About to call run_bmi_server...
Logic error caught: basic_string: construction from null is not valid
This suggests a null pointer passed to string constructor
```
Have updated the changes in dockerfile for the test routine on the dockerfile and resultant image run https://github.com/nishadhka/hype-bmi/commit/bb85b7e29767f6b9bd07f7c0dbb289432d2936bd

**Comment by BSchilperoort** [comment]
I don't think Claude has sufficient BMI use cases in its training data to be able to debug this fully.

```
        // Test initialize with current directory
        std::cout << "Testing BMI initialize with current directory..." << std::endl;
        int result = model->initialize(".");
```

Initialization will fail here as you need to pass the path to a config file, not a directory. Apparently the HYPE bmi also allows for a directory but this would still require the right config and/or input data. It probably goes wrong somewhere here; https://github.com/nishadhka/hype-bmi/blob/bb85b7e29767f6b9bd07f7c0dbb289432d2936bd/src/hypelib.f90#L246C1-L246C13

I would not expect that you need to rewrite some of these files, or port things to hype 5.6.2. If you would be able to pin the right grpc & grpc4bmi versions, it should be possible to build a working container again. To test the model you would need valid input data.

**Comment by BSchilperoort** [comment]
Also, based on the last change to the dockerfile you should be using [grpc 1.19.1](https://github.com/grpc/grpc/releases/tag/v1.19.1).

**Comment by BSchilperoort** [comment]
> Also, based on the last change to the dockerfile you should be using [grpc 1.19.1](https://github.com/grpc/grpc/releases/tag/v1.19.1).

which seems to not work anymore as the submodules are broken because those repositories don't exist anymore 🙄

**Comment by nishadhka** [comment]
Understood, checking at hype 5.6.2 and code update needed. I am feeling can the effort be channeled to work on the latest version of the Hype model instead. 
Cheked on these two repos, will see any work around, otherwise will take pause. Thank you for the help. 
https://github.com/csdms/bmi-docker
https://github.com/csdms/grpc4bmi-docker

**Comment by BSchilperoort** [comment]
Yeah I think working with latest versions everywhere might be the way to go. However this is sadly not easy... (which is why it's been broken so long).

Maybe start with getting (standalone) Hype to run, if you haven't already? And then add the other parts
